### PR TITLE
[WIP] Moving towards .NET Standard/.NET Core compatibility

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
**!Important!** This PR is not about introducing NET CLI or xproj/project.json in the source code. It makes the source code easier to introduce .NET Platform compatibility in the future.

I used some ideas from this thread https://github.com/akkadotnet/akka.net/issues/992

Right now the sources match NetStandard 1.5, but it possible to achieve 1.3 in the future

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests/ClusterShardingSpec.cs [hunk 1]
```diff
@@ -295,7 +295,7 @@ private void CreateCoordinator()
 
             foreach (var typeName in typeNames)
             {
-                var rebalanceEnabled = string.Equals(typeName, "rebalancing", StringComparison.InvariantCultureIgnoreCase);
+                var rebalanceEnabled = string.Equals(typeName, "rebalancing", StringComparison.OrdinalIgnoreCase);
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests/ClusterShardingSpec.cs]: MS recommends to replace InvariantCultureIgnoreCase to OrdinalIgnoreCase in many cases

File: src/core/Akka/Dispatch/AbstractDispatcher.cs [hunk 2]
```diff
@@ -191,12 +191,16 @@ private static DedicatedThreadPoolSettings ConfigureSettings(Config config)
     /// </summary>
     internal sealed class ThreadPoolExecutorServiceFactory : ExecutorServiceConfigurator
     {
+#if UNSAFE_THREADING
```
Comment (alexvaluyskiy) [inline src/core/Akka/Dispatch/AbstractDispatcher.cs]: NetCore doesn't have AppDomains

File: src/core/Akka/Configuration/ConfigurationFactory.cs [hunk 3]
```diff
@@ -62,10 +61,12 @@ public static Config ParseString(string hocon)
...
         {
-            var section = (AkkaConfigurationSection)ConfigurationManager.GetSection("akka") ?? new AkkaConfigurationSection();
-            var config = section.AkkaConfig;
-
-            return config;
+#if CONFIGURATION
+            var section = (AkkaConfigurationSection)System.Configuration.ConfigurationManager.GetSection("akka") ?? new AkkaConfigurationSection();
+            return section.AkkaConfig;
+#else
+            return ConfigurationFactory.Empty;
```
Comment (alexvaluyskiy) [inline src/core/Akka/Configuration/ConfigurationFactory.cs]: For compatibility reasons ConfigurationFactory.Load just returns empty config on NetCor

File: src/core/Akka/Dispatch/ThreadPoolBuilder.cs [hunk 4]
```diff
@@ -28,19 +28,21 @@ internal static class DedicatedThreadPoolConfigHelpers
 
         internal static ThreadType ConfigureThreadType(string threadType)
         {
-            return string.Compare(threadType, ThreadType.Foreground.ToString(), StringComparison.InvariantCultureIgnoreCase) == 0 ?
+            return string.Compare(threadType, ThreadType.Foreground.ToString(), StringComparison.OrdinalIgnoreCase) == 0 ?
                 ThreadType.Foreground : ThreadType.Background;
         }
 
+#if UNSAFE_THREADING
         internal static ApartmentState GetApartmentState(Config cfg)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Dispatch/ThreadPoolBuilder.cs]: AppartmentState

File: src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs [hunk 5]
```diff
@@ -304,15 +325,21 @@ public PoolWorker(DedicatedThreadPool pool, int workerId)
                 _pool = pool;
                 _threadExit = new TaskCompletionSource<object>();
 
+#if UNSAFE_THREADING
                 var thread = new Thread(RunThread, pool.Settings.ThreadMaxStackSize);
```
Comment (alexvaluyskiy) [inline src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs]: NetCore version doesn't have a parameter for ThreadMaxStackSize

File: src/core/Akka/Dispatch/Mailbox.cs [hunk 6]
```diff
@@ -338,10 +338,12 @@ private void ProcessAllSystemMessages()
                 msg.Unlink();
                 DebugPrint("{0} processing system message {1} with {2}", Actor.Self, msg, string.Join(",", Actor.GetChildren()));
                 // we know here that SystemInvoke ensures that only "fatal" exceptions get rethrown
+#if UNSAFE_THREADING
```
Comment (alexvaluyskiy) [inline src/core/Akka/Dispatch/Mailbox.cs]: Should I ignore these exceptions on NetCore?

File: src/core/Akka/Configuration/Hocon/HoconTokenizer.cs [hunk 7]
```diff
@@ -559,7 +559,7 @@ private string PullEscapeSequence()
                 case 'u':
                     string hex = "0x" + Take(4);
                     int j = Convert.ToInt32(hex, 16);
-                    return ((char) j).ToString(CultureInfo.InvariantCulture);
+                    return ((char) j).ToString();
```
Comment (alexvaluyskiy) [inline src/core/Akka/Configuration/Hocon/HoconTokenizer.cs]: I'm not sure that I did the right conversion

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt [hunk 8]
```diff
@@ -371,7 +371,7 @@ namespace Akka.Actor
             where T :  class, Akka.Actor.IExtension
             where TI : Akka.Actor.IExtensionId { }
     }
-    public sealed class Address : Akka.Util.ISurrogated, System.ICloneable, System.IEquatable<Akka.Actor.Address>
```
Comment (alexvaluyskiy) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCore.approved.txt]: Should I remove ICloneable interface or use compilation directives

File: src/core/Akka/Actor/Address.cs [hunk 9]
```diff
@@ -22,7 +22,7 @@ namespace Akka.Actor
     /// for example a remote transport would want to associate additional
     /// information with an address, then this must be done externally.
     /// </summary>
-    public sealed class Address : ICloneable, IEquatable<Address>, ISurrogated
```
Comment (alexvaluyskiy) [inline src/core/Akka/Actor/Address.cs]: Should I remove ICloneable interface or use compilation directives?

**Comment by alexvaluyskiy** [comment]
@akkadotnet/developers So, the main question. What should we do with BinarySerialization in exceptions, and with SerializableAttribute? We could just remove it, as many projects did. Or we can add compilation directives, as I did here

**Comment by Aaronontheweb** [comment]
paging @JeffCyr - questions about unsafe threading on .NET Core :p

File: src/core/Akka/Dispatch/Mailbox.cs (hunk 6 above)
Comment (JeffCyr) [inline src/core/Akka/Dispatch/Mailbox.cs]: As far as I know the ThreadInterruptedException and ThreadAbortException handlers here are useless.

Actor.SystemInvoke isn't blocking, so ThreadInterruptedException can't be thrown.

In the ThreadAbortException handler, Thread.ResetAbort() isn't called so the exception will be [re-thrown automatically](https://msdn.microsoft.com/en-us/library/system.threading.threadabortexception(v=vs.110).aspx#Anchor_4).

ThreadAbort resistance is very hard to get right and I don't think it is properly implemented here even if Thread.ResetAbort was called, I would advise to simply remove thoses exceptions handlers and never abort threads.

File: src/core/Akka.Streams/Implementation/Fusing/GraphInterpreter.cs [hunk 10]
```diff
@@ -165,7 +165,7 @@ public static GraphInterpreter Current
             get
             {
                 if (CurrentInterpreter.Value[0] == null)
-                    throw new ApplicationException("Something went terribly wrong!");
+                    throw new InvalidOperationException("Something went terribly wrong!");
```
Comment (alexvaluyskiy) [inline src/core/Akka.Streams/Implementation/Fusing/GraphInterpreter.cs]: NetCore doesn't have ApplicatioNException, changed to InvalidOperationException

File: src/core/Akka.Tests/Configuration/ConfigurationSpec.cs [hunk 11]
```diff
@@ -65,15 +64,17 @@ public void The_default_configuration_file_contain_all_configuration_properties(
             settings.SchedulerClass.ShouldBe(typeof(DedicatedThreadScheduler).FullName);
         }
 
+#if CONFIGURATION
         [Fact]
         public void Deserializes_hocon_configuration_from_net_config_file()
```
Comment (alexvaluyskiy) [inline src/core/Akka.Tests/Configuration/ConfigurationSpec.cs]: Skipped this test for NetCore

File: src/core/Akka.Tests/TestUtils/Comparable.cs [hunk 12]
```diff
@@ -23,7 +23,7 @@ public override int GetHashCode()
 
         public override string ToString()
         {
-            var res = fastJSON.JSON.Instance.ToJSON(this);
+            var res = Newtonsoft.Json.JsonConvert.SerializeObject(this);
```
Comment (alexvaluyskiy) [inline src/core/Akka.Tests/TestUtils/Comparable.cs]: Replaced fastJson to NetwonsoftJson

File: src/core/Akka/Dispatch/ThreadPoolBuilder.cs (hunk 4 above)
Comment (alexvaluyskiy) [inline src/core/Akka/Dispatch/ThreadPoolBuilder.cs]: @JeffCyr could we live without AppartmentState?

File: src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs (hunk 5 above)
Comment (alexvaluyskiy) [inline src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs]: @JeffCyr, what do you think about it?

Comment (JeffCyr) [inline src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs]: I think it's not configurable from akka anyway, LGTM

File: src/core/Akka/Dispatch/Mailbox.cs (hunk 6 above)
Comment (alexvaluyskiy) [inline src/core/Akka/Dispatch/Mailbox.cs]: @Aaronontheweb could I remove these two handlers?

File: src/core/Akka/Configuration/Hocon/HoconTokenizer.cs (hunk 7 above)
Comment (alexvaluyskiy) [inline src/core/Akka/Configuration/Hocon/HoconTokenizer.cs]: The code from ReferenceSource
```C#
[Pure]
public String ToString(IFormatProvider provider) { 
    Contract.Ensures(Contract.Result<String>() != null);
    return Char.ToString(m_value); 
} 
```
They don't use this parameter

File: src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs [hunk 13]
```diff
@@ -69,7 +72,23 @@ internal class DedicatedThreadPoolSettings
             if (numThreads <= 0)
                 throw new ArgumentOutOfRangeException("numThreads", string.Format("numThreads must be at least 1. Was {0}", numThreads));
         }
+#else 
+        public DedicatedThreadPoolSettings(int numThreads, string name = null, TimeSpan? deadlockTimeout = null)
```
Comment (alexvaluyskiy) [inline src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs]: Added a different constructor for NetCore version, without ApartmentState

**Comment by adamhathcock** [comment]
Say it ain't so!

**Comment by alexvaluyskiy** [comment]
We should create a feature branch here instead of this PR

**Comment by Aaronontheweb** [comment]
@alexvaluyskiy at the next contributor's meeting, let's sort that out for how we want to do this incrementally.

**Comment by alexvaluyskiy** [comment]
@Aaronontheweb Ok.
Right now I'm working on it in my local branch 
https://github.com/alexvaluyskiy/akka.net/commits/netcorecodefixes

**Comment by sdanishali** [comment]
Does AKKA.Net now work with ASP.Net Core?

http://fizzylogic.nl/2016/05/22/Using-Akka-NET-from-your-ASP-NET-core-application/
https://github.com/wmeints/akka-aspnet-core

**Comment by wmeints** [comment]
@sdanishali No it does not, it only supports full .NET framework. This is mentioned in the comments of the post :-) Sorry dude, it's going to take some time before Akka.NET supports the .NET standard way of doing things. Until then use net45 or higher instead.

**Pull request closed without merging**